plugins/hls-rename-plugin/src/Ide/Plugin/Rename.hs

```diff
@@ -0,0 +1,42 @@
+{-# LANGUAGE DataKinds             #-}
+{-# LANGUAGE DuplicateRecordFields #-}
+
+module Ide.Plugin.Rename (descriptor) where
+
+import           Data.Function
+import           Data.HashMap.Internal               hiding (map)
+import           Data.List
+import qualified Data.Text                           as T
+import           Development.IDE                     hiding (pluginHandlers)
+import           Development.IDE.LSP.HoverDefinition
+import           Ide.Types
+import           Language.LSP.Server
+import           Language.LSP.Types
+
+descriptor :: PluginId -> PluginDescriptor IdeState
+descriptor pluginId = (defaultPluginDescriptor pluginId) {
+    pluginHandlers = mkPluginHandler STextDocumentRename renameProvider
+}
+
+renameProvider :: PluginMethodHandler IdeState TextDocumentRename
+renameProvider state _pluginId (RenameParams tdi pos _progToken name) = do
+    locs <- getTextEdits state tdi pos name
+    return $ Right (WorkspaceEdit {
+        _changes = Just (fromList locs),
+        _documentChanges = Nothing,
+        _changeAnnotations = Nothing
+    })
+
+getTextEdits :: IdeState
+    -> TextDocumentIdentifier
+    -> Position
+    -> T.Text
+    -> LspT c IO [(Uri, List TextEdit)]
+getTextEdits state tdi pos name = do
+    mbLocs <- references state $ ReferenceParams tdi pos Nothing Nothing (ReferenceContext False)
```

Comment:
References are reported for things like qualified names (the entire token `Data.Map.insert` is a reference to `insert`), are implicitly added on to record wildcards(the `{..}` in `Foo{..}` references all the fields of `Foo`) and perhaps more edge cases I'm missing. These will all need to be handled carefully.

Reply:
Currently, my solution to the qualified name problem is to use the length of the occurrence name to set the correct length of the `textEdit`https://github.com/haskell/haskell-language-server/blob/a80bf10ac07f2d804ce281831c9ebe3c163f9a58/plugins/hls-rename-plugin/src/Ide/Plugin/Rename.hs#L94-L98
I hadn't thought about record wildcards yet thought, so I'll make sure to add a test for that shortly.
Thanks 🙏